### PR TITLE
More flexible calendar implementation [clean]

### PR DESCRIPTION
**Pull request by warren-oneill**
My PR #523 became a bit unwieldy so I decided to make a new PR from a fresh branch. 
@ehebert, @twiecki, @jfkirk

### CODE REVIEW

File: tests/test_tradingcalendar.py [hunk 1]
```diff
@@ -26,7 +26,6 @@
 
 class TestTradingCalendar(TestCase):
 
-    @nottest
```
Comment (jfkirk) [inline tests/test_tradingcalendar.py]: Why are we disabling this test?

File: zipline/finance/trading.py [hunk 2]
```diff
@@ -93,13 +92,34 @@ def __init__(
         max_date=None,
         env_trading_calendar=tradingcalendar
     ):
+        # `tc_tdays` is short for "trading calendar trading days"
+        # `tc_tday` is short for "trading calendar trading day"
+        tc_tdays = env_trading_calendar.trading_days
+        tc_tday = env_trading_calendar.trading_day
```
Comment (jfkirk) [inline zipline/finance/trading.py]: What is the purpose of tc_tday? Is it a definition of the trading hours?

Comment (warren-oneill) [inline zipline/finance/trading.py]: trading_days is a pandas date range and trading_day is declared using [Custom Business Day](http://pandas.pydata.org/pandas-docs/version/0.12.0/timeseries.html#custom-business-days-experimental)  as follows
```
trading_day = pd.tseries.offsets.CDay(holidays=non_trading_days)
```

File: tests/test_tradingcalendar.py (hunk 1 above)
Comment (warren-oneill) [inline tests/test_tradingcalendar.py]: I'm not sure why it was disable but the test for BMF, TSE and LSE are also commented out but I'm pretty sure all but TSE will fail.

File: zipline/finance/trading.py (hunk 2 above)
Comment (jfkirk) [inline zipline/finance/trading.py]: Ah, got it, thanks. 

The purpose of adding the tc_tday field, then, is to expose this Custom Business Day object through the TradingEnvironment's 'trading_day' field? Where would this field be read?

Comment (ehebert) [inline zipline/finance/trading.py]: There may have been some improvements to the how to define holidays in Pandas, but I haven't caught up completely on it.

https://github.com/pydata/pandas/issues/7070

File: tests/test_tradingcalendar.py (hunk 1 above)
Comment (jfkirk) [inline tests/test_tradingcalendar.py]: @ehebert @llllllllll Do either of you know why this test was disabled?

File: zipline/finance/trading.py (hunk 2 above)
Comment (warren-oneill) [inline zipline/finance/trading.py]: tc_tday is used by loader()  in TradingEnvironment. Previously trading_day and trading_days for NYSE were hardcoded in loader.py but I suggested to pass them as a variables to loader().

File: tests/test_tradingcalendar.py (hunk 1 above)
Comment (ehebert) [inline tests/test_tradingcalendar.py]: I can't remember the circumstances exactly, but IIRC we had to overhaul the calendar for some reason, and only had the time to port the U.S. Equity calendar.

(Hoping the more flexible calendar will help not having to make that tradeoff again.)

Comment (warren-oneill) [inline tests/test_tradingcalendar.py]: Hopefully for BMF and TSE everything will be green but currently LSE only has days defined and ignores trading times (e.g. it has no get_early_closes, get_open_and_closes....).

File: zipline/finance/trading.py (hunk 2 above)
Comment (jfkirk) [inline zipline/finance/trading.py]: Ok, thanks. The purpose is clear to me now.

**Comment by jfkirk** [comment]
This PR looks good - we are now going to run it through our internal validation gamut and let you know when we are ready to merge.

**Comment by warren-oneill** [comment]
Good to hear, thanks!

File: zipline/data/loader.py [hunk 3]
```diff
@@ -31,10 +31,6 @@
 from . import benchmarks
 from . benchmarks import get_benchmark_returns
 
-from zipline.utils.tradingcalendar import (
-    trading_day,
-    trading_days
-)
```
Comment (jfkirk) [inline zipline/data/loader.py]: Let's leave this in and use them as default arg values for load_market_data

File: zipline/data/loader.py [hunk 4]
```diff
@@ -154,7 +150,7 @@ def get_benchmark_filename(symbol):
     return "%s_benchmark.csv" % symbol
 
 
-def load_market_data(bm_symbol='^GSPC'):
+def load_market_data(trading_day, trading_days, bm_symbol='^GSPC'):
```
Comment (jfkirk) [inline zipline/data/loader.py]: Let's add default values for trading_day and trading_days (the values from zipline.utils.tradingcalendar) to support backwards-compatability.

**Comment by jfkirk** [comment]
@warren-oneill I added some notes for backwards compatibility. This will satisfy our internal requirements.

File: zipline/data/loader.py [hunk 5]
```diff
@@ -31,10 +31,8 @@
 from . import benchmarks
 from . benchmarks import get_benchmark_returns
 
-from zipline.utils.tradingcalendar import (
-    trading_day,
-    trading_days
-)
+from zipline.utils.tradingcalendar import trading_day as trading_day_nyse
+from zipline.utils.tradingcalendar import trading_days as trading_days_nyse
```
Comment (warren-oneill) [inline zipline/data/loader.py]: I decided to append "nyse" to the names to make their usage clearer. Tbh I feel tradingcalendar should also be named tradingcalendar_nyse.

File: zipline/finance/trading.py [hunk 6]
```diff
@@ -93,13 +92,34 @@ def __init__(
...
+
+        if max_date:
+            self.trading_days = tc_tdays[tc_tdays <= max_date].copy()
+            self.trading_day = tc_tday[tc_tday <= max_date].copy()
```
Comment (jfkirk) [inline zipline/finance/trading.py]: This throws an error for me:

 File ".../zipline/finance/trading.py", line 102, in __init__
    self.trading_day = tc_tday[tc_tday <= max_date].copy()
TypeError: can't compare datetime.datetime to CustomBusinessDay

**Comment by ssanderson** [comment]
@jfkirk @ehebert what's the status on this?  I think I'm about to re-write a bunch of tradingcalendar.py to do ~a second's worth of work at import time, since that's having a substantial user-facing impact in the research environment.

**Comment by ehebert** [comment]
@ssanderson I believe this is pending a run of the internal algo test harness.

**Comment by warren-oneill** [comment]
@ssanderson @ehebert James put it through the internal test harness yesterday. I just need to rebase off master and then I think its ready to be merged.

**Comment by warren-oneill** [comment]
I ran Scrutinizer from my fork and it was green. Is it possible to rerun it?

**Comment by jfkirk** [comment]
Closing this PR in deference to this: https://github.com/quantopian/zipline/pull/560

**Pull request closed without merging**